File: frontend/src/utils/passwordValidator.ts
/**
 * Password validation utilities for the frontend
 */

export interface PasswordRequirement {
  text: string;
  isValid: boolean;
}

export interface PasswordValidationResult {
  isValid: boolean;
  requirements: PasswordRequirement[];
  score: number; // 0-100 strength score
}

export class PasswordValidator {
  private static readonly MIN_LENGTH = 8;
  private static readonly SPECIAL_CHARACTERS = "!@#$%^&*()_+-=[]{}|;:,.<>?";

  private static readonly COMMON_PASSWORDS = new Set([
    "password",
    "123456",
    "password123",
    "admin",
    "qwerty",
    "letmein",
    "welcome",
    "monkey",
    "1234567890",
    "abc123",
    "123456789",
    "password1",
    "iloveyou",
    "123123",
    "admin123",
    "qwerty123",
    "password12",
    "welcome123",
    "master",
    "hello",
    "login",
    "guest",
    "test",
    "user",
    "root",
    "toor",
  ]);

  /**
   * Validate password and return detailed results
   */
  static validatePassword(password: string): PasswordValidationResult {
    const requirements: PasswordRequirement[] = [
      {
        text: `At least ${this.MIN_LENGTH} characters long`,
        isValid: password.length >= this.MIN_LENGTH,
      },
      {
        text: "At least one uppercase letter (A-Z)",
        isValid: /[A-Z]/.test(password),
      },
      {
        text: "At least one lowercase letter (a-z)",
        isValid: /[a-z]/.test(password),
      },
      {
        text: "At least one digit (0-9)",
        isValid: /\d/.test(password),
      },
      {
        text: `At least one special character (${this.SPECIAL_CHARACTERS})`,
        isValid: new RegExp(
          `[${this.escapeRegExp(this.SPECIAL_CHARACTERS)}]`
        ).test(password),
      },
      {
        text: "Not a common password",
        isValid: !this.COMMON_PASSWORDS.has(password.toLowerCase()),
      },
      {
        text: "No more than 3 consecutive identical characters",
        isValid: !/(.)\1{3,}/.test(password),
      },
      {
        text: "No simple sequences (like '1234' or 'abcd')",
        isValid: !this.containsSimpleSequence(password),
      },
    ];

    const validCount = requirements.filter((req) => req.isValid).length;
    const isValid = validCount === requirements.length;
    const score = Math.round((validCount / requirements.length) * 100);

    return {
      isValid,
      requirements,
      score,
    };
  }

  /**
   * Get password strength level based on score
   */
  static getStrengthLevel(score: number): "weak" | "fair" | "good" | "strong" {
    if (score < 40) return "weak";
    if (score < 60) return "fair";
    if (score < 80) return "good";
    return "strong";
  }

  /**
   * Get color for password strength indicator
   */
  static getStrengthColor(score: number): string {
    if (score < 40) return "#ef4444"; // red
    if (score < 60) return "#f59e0b"; // yellow
    if (score < 80) return "#3b82f6"; // blue
    return "#10b981"; // green
  }

  private static escapeRegExp(string: string): string {
    return string.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  }

  private static containsSimpleSequence(password: string): boolean {
    const passwordLower = password.toLowerCase();

    const sequences = [
      "0123456789",
      "abcdefghijklmnopqrstuvwxyz",
      "qwertyuiopasdfghjklzxcvbnm", // QWERTY keyboard layout
    ];

    for (const sequence of sequences) {
      for (let i = 0; i <= sequence.length - 4; i++) {
        const subseq = sequence.substring(i, i + 4);
        const reverseSubseq = subseq.split("").reverse().join("");

        if (
          passwordLower.includes(subseq) ||
          passwordLower.includes(reverseSubseq)
        ) {
          return true;
        }
      }
    }

    return false;
  }
}

File: backend/app/password_validator.py
"""
Password validation utilities for enforcing strong password policies.
"""

import re
from typing import List, Tuple


class PasswordValidationError(Exception):
    """Custom exception for password validation errors."""

    def __init__(self, messages: List[str]):
        self.messages = messages
        super().__init__("; ".join(messages))


class PasswordValidator:
    """
    Password validator that enforces strong password requirements.

    Requirements:
    - At least 8 characters long
    - At least one uppercase letter
    - At least one lowercase letter
    - At least one digit
    - At least one special character (!@#$%^&*()_+-=[]{}|;:,.<>?)
    - No common passwords
    - No more than 3 consecutive identical characters
    - No simple sequences (like '1234' or 'abcd')
    - Cannot contain the username
    """

    MIN_LENGTH = 8
    SPECIAL_CHARACTERS = "!@#$%^&*()_+-=[]{}|;:,.<>?"

    # Common passwords to reject
    COMMON_PASSWORDS = {
        "password",
        "123456",
        "password123",
        "admin",
        "qwerty",
        "letmein",
        "welcome",
        "monkey",
        "1234567890",
        "abc123",
        "123456789",
        "password1",
        "iloveyou",
        "123123",
        "admin123",
        "qwerty123",
        "password12",
        "welcome123",
        "master",
        "hello",
        "login",
        "guest",
        "test",
        "user",
        "root",
        "toor",
    }

    @classmethod
    def validate_password(
        cls, password: str, username: str = ""
    ) -> Tuple[bool, List[str]]:
        """
        Validate a password against all requirements.

        Args:
            password: The password to validate
            username: The username to check against (password cannot contain username)

        Returns:
            Tuple of (is_valid, error_messages)
        """
        errors = []

        # Check minimum length
        if len(password) < cls.MIN_LENGTH:
            errors.append(f"Password must be at least {cls.MIN_LENGTH} characters long")

        # Check for uppercase letter
        if not re.search(r"[A-Z]", password):
            errors.append("Password must contain at least one uppercase letter")

        # Check for lowercase letter
        if not re.search(r"[a-z]", password):
            errors.append("Password must contain at least one lowercase letter")

        # Check for digit
        if not re.search(r"\d", password):
            errors.append("Password must contain at least one digit")

        # Check for special character
        special_char_pattern = f"[{re.escape(cls.SPECIAL_CHARACTERS)}]"
        if not re.search(special_char_pattern, password):
            errors.append(
                f"Password must contain at least one special character ({cls.SPECIAL_CHARACTERS})"
            )

        # Check against common passwords
        if password.lower() in cls.COMMON_PASSWORDS:
            errors.append("Password is too common and easily guessable")

        # Check for repeated characters (more than 3 consecutive)
        if re.search(r"(.)\1{3,}", password):
            errors.append(
                "Password cannot contain more than 3 consecutive identical characters"
            )

        # Check for simple sequences
        if cls._contains_simple_sequence(password):
            errors.append(
                "Password cannot contain simple sequences like '1234' or 'abcd'"
            )

        # Check if password contains username
        if username and len(username) >= 3 and username.lower() in password.lower():
            errors.append("Password cannot contain the username")

        return len(errors) == 0, errors

    @classmethod
    def _contains_simple_sequence(cls, password: str) -> bool:
        """Check if password contains simple sequences."""
        password_lower = password.lower()

        # Check for ascending sequences
        sequences = [
            "0123456789",
            "abcdefghijklmnopqrstuvwxyz",
            "qwertyuiopasdfghjklzxcvbnm",  # QWERTY keyboard layout
        ]

        for sequence in sequences:
            for i in range(len(sequence) - 3):
                if sequence[i : i + 4] in password_lower:
                    return True
                # Check reverse sequences too
                if sequence[i : i + 4][::-1] in password_lower:
                    return True

        return False

    @classmethod
    def validate_and_raise(cls, password: str, username: str = "") -> None:
        """
        Validate password and raise PasswordValidationError if invalid.

        Args:
            password: The password to validate
            username: The username to check against (password cannot contain username)

        Raises:
            PasswordValidationError: If password doesn't meet requirements
        """
        is_valid, errors = cls.validate_password(password, username)
        if not is_valid:
            raise PasswordValidationError(errors)

    @classmethod
    def get_requirements_text(cls) -> str:
        """Get a human-readable description of password requirements."""
        return (
            f"Password must meet the following requirements:\n"
            f"• At least {cls.MIN_LENGTH} characters long\n"
            f"• At least one uppercase letter (A-Z)\n"
            f"• At least one lowercase letter (a-z)\n"
            f"• At least one digit (0-9)\n"
            f"• At least one special character ({cls.SPECIAL_CHARACTERS})\n"
            f"• Cannot be a common password\n"
            f"• Cannot contain more than 3 consecutive identical characters\n"
            f"• Cannot contain simple sequences (like '1234' or 'abcd')\n"
            f"• Cannot contain the username"
        )
